prow/spyglass/gcsartifact_test.go

```diff
@@ -0,0 +1,248 @@
+		artifact := NewGCSArtifact(context.Background(), &fakeArtifactHandle{
+			contents: tc.contents,
+			oAttrs: &storage.ObjectAttrs{
+				Bucket:          "foo-bucket",
+				Name:            "build-log.txt",
+				Size:            int64(len(tc.contents)),
+				ContentEncoding: tc.encoding,
+			},
+		}, "", "build-log.txt", 500e6)
+
+		actualBytes, err := artifact.ReadAtMost(tc.n)
+		if err != nil && !tc.expectErr {
+			t.Fatalf("Test %s failed with err:\n%s", tc.name, err)
+		}
+		if err == nil && tc.expectErr {
+			t.Errorf("Test %s did not produce error when expected", tc.name)
+		}
+		if !bytes.Equal(actualBytes, tc.expected) {
+			t.Errorf("Test %s failed.\nExpected: %s\nActual: %s", tc.name, tc.expected, actualBytes)
+		}
+	}
+}
+
+// Tests reading all data from files in GCS
+func TestGCSReadAll(t *testing.T) {
+	testCases := []struct {
+		name        string
+		sizeLimit   int64
+		contents    []byte
+		expectedErr error
+		expected    []byte
+	}{
+		{
+			name:        "ReadAll example build log",
+			expectedErr: nil,
+			contents:    []byte("Oh wow\nlogs\nthis is\ncrazy"),
+			sizeLimit:   500e6,
+			expected:    []byte("Oh wow\nlogs\nthis is\ncrazy"),
+		},
+		{
+			name:        "ReadAll example too large build log",
+			sizeLimit:   20,
+			contents:    []byte("Oh wow\nlogs\nthis is\ncrazy"),
+			expectedErr: viewers.ErrFileTooLarge,
+			expected:    nil,
+		},
+	}
+	for _, tc := range testCases {
+		artifact := NewGCSArtifact(context.Background(), &fakeArtifactHandle{
+			contents: tc.contents,
+			oAttrs: &storage.ObjectAttrs{
+				Bucket: "foo-bucket",
+				Name:   "build-log.txt",
+				Size:   int64(len(tc.contents)),
+			},
+		}, "", "build-log.txt", tc.sizeLimit)
+
+		actualBytes, err := artifact.ReadAll()
+		if err != nil {
+			if err != tc.expectedErr {
+				t.Errorf("Test %s got err %v, but expected err: %v", tc.name, err, tc.expectedErr)
+			}
+		}
+		if !bytes.Equal(actualBytes, tc.expected) {
+			t.Errorf("Test %s failed.\nExpected: %s\nActual: %s", tc.name, tc.expected, actualBytes)
+		}
+	}
+}
+
+func TestGCSSize(t *testing.T) {
+	startedHandle := &fakeArtifactHandle{
+		contents: []byte(`{
+						  "node": "gke-prow-default-pool-3c8994a8-qfhg", 
+						  "repo-version": "v1.12.0-alpha.0.985+e6f64d0a79243c", 
+						  "timestamp": 1528742858, 
+						  "repos": {
+						    "k8s.io/kubernetes": "master", 
+						    "k8s.io/release": "master"
+						  }, 
+						  "version": "v1.12.0-alpha.0.985+e6f64d0a79243c", 
+						  "metadata": {
+						    "pod": "cbc53d8e-6da7-11e8-a4ff-0a580a6c0269"
+						  }
+						}`),
+		oAttrs: &storage.ObjectAttrs{
+			Bucket: "foo-bucket",
+			Name:   "started.json",
+			Size:   418,
+		},
+	}
+	nonexistentHandle := &gcsArtifactHandle{fakeGCSBucket.Object("logs/example-ci-run/404/started.json")}
+	testCases := []struct {
+		name      string
+		handle    artifactHandle
+		expected  int64
+		expectErr bool
+	}{
+		{
+			name:   "Test size simple",
+			handle: startedHandle,
```

Comment:
Looks like startedHandle is only used in one location, so please move it into here